From 1e9a2e1fc0f37b1f549f21636ca40bc0ec3b72f5 Mon Sep 17 00:00:00 2001
From: MutsukiC <mutsuki@aosc.io>
Date: Sat, 16 May 2026 13:06:04 +0800
Subject: [PATCH 4/5] Adapt ZXing 3.0 API Changes

Backported these upstream changes:
- Port away from deprecated ZXing API
- Adapt to ZXing 3.0 BarcodeFormats API change
---
 src/lib/barcodedecoder.cpp | 25 ++++++++++++++++++++-----
 1 file changed, 20 insertions(+), 5 deletions(-)

diff --git a/src/lib/barcodedecoder.cpp b/src/lib/barcodedecoder.cpp
index 2892665b..ac2058d3 100644
--- a/src/lib/barcodedecoder.cpp
+++ b/src/lib/barcodedecoder.cpp
@@ -152,9 +152,9 @@ struct {
 #endif
 };
 
-static auto typeToFormats(BarcodeDecoder::BarcodeTypes types)
+static ZXing::BarcodeFormats typeToFormats(BarcodeDecoder::BarcodeTypes types)
 {
-#if ZXING_VERSION >= QT_VERSION_CHECK(1, 1, 0)
+#if ZXING_VERSION < QT_VERSION_CHECK(3, 0, 0)
     ZXing::BarcodeFormats formats;
 #else
     std::vector<ZXing::BarcodeFormat> formats;
@@ -162,7 +162,7 @@ static auto typeToFormats(BarcodeDecoder::BarcodeTypes types)
 
     for (auto i : zxing_format_map) {
         if (types & i.type) {
-#if ZXING_VERSION >= QT_VERSION_CHECK(1, 1, 0)
+#if ZXING_VERSION < QT_VERSION_CHECK(3, 0, 0)
             formats |= i.zxingType;
 #else
             formats.push_back(i.zxingType);
@@ -189,7 +189,11 @@ static ZXing::ImageFormat zxingImageFormat(QImage::Format format)
         case QImage::Format_ARGB32:
         case QImage::Format_RGB32:
 #if Q_BYTE_ORDER == Q_LITTLE_ENDIAN
-            return ZXing::ImageFormat::BGRX;
+#if ZXING_VERSION >= QT_VERSION_CHECK(2, 3, 0)
+            return ZXing::ImageFormat::RGBA;
+#else
+            return ZXing::ImageFormat::RGBX;
+#endif
 #else
             return ZXing::ImageFormat::XRGB;
 #endif
@@ -197,7 +201,11 @@ static ZXing::ImageFormat zxingImageFormat(QImage::Format format)
             return ZXing::ImageFormat::RGB;
         case QImage::Format_RGBX8888:
         case QImage::Format_RGBA8888:
+#if ZXING_VERSION >= QT_VERSION_CHECK(2, 3, 0)
+            return ZXing::ImageFormat::RGBA;
+#else
             return ZXing::ImageFormat::RGBX;
+#endif
         case QImage::Format_Grayscale8:
             return ZXing::ImageFormat::Lum;
         default:
@@ -205,8 +213,11 @@ static ZXing::ImageFormat zxingImageFormat(QImage::Format format)
     }
     Q_UNREACHABLE();
 }
-
+#if ZXING_VERSION >= QT_VERSION_CHECK(3, 0, 0)
+static ZXing::Barcode zxingReadBarcode(const QImage &img, const ZXing::ReaderOptions &hints)
+#else
 static ZXing::Result zxingReadBarcode(const QImage &img, const ZXing::DecodeHints &hints)
+#endif
 {
     return ZXing::ReadBarcode({img.bits(), img.width(), img.height(), zxingImageFormat(img.format()), static_cast<int>(img.bytesPerLine())}, hints);
 }
@@ -214,7 +225,11 @@ static ZXing::Result zxingReadBarcode(const QImage &img, const ZXing::DecodeHint
 
 void BarcodeDecoder::decodeZxing(const QImage &img, BarcodeDecoder::BarcodeTypes format, BarcodeDecoder::Result &result) const
 {
+#if ZXING_VERSION >= QT_VERSION_CHECK(3, 0, 0)
+    ZXing::ReaderOptions hints;
+#else
     ZXing::DecodeHints hints;
+#endif
 #if ZXING_VERSION >= QT_VERSION_CHECK(1, 1, 0)
     hints.setFormats(typeToFormats(format));
 #else
-- 
2.52.0

